File: acorn/permissions.py
"""Permission system — three modes with session-scoped allow rules."""

import re

# Tools that never need approval
ALWAYS_SAFE = {'read_file', 'glob', 'grep'}

# Patterns that ALWAYS require approval even in auto mode
DANGEROUS_PATTERNS = [
    # Unix
    r'\brm\s+(-rf?|--recursive)', r'\brm\s+/', r'rmdir\s+/',
    r'\bmkfs\b', r'>\s*/dev/', r'dd\s+if=',
    r'chmod\s+(-R\s+)?777', r'chown\s+-R\s+.*/',
    r'\bgit\s+push\s+.*--force', r'\bgit\s+reset\s+--hard',
    r'\bdrop\s+table\b', r'\bdrop\s+database\b',
    r'\btruncate\s+table\b',
    r'\bmkfs\.\w+\b', r'\bfdisk\b', r'\bparted\b',
    r':()\{', r'curl.*\|\s*(ba)?sh', r'wget.*\|\s*(ba)?sh',
    r'\bkill\s+-9\b',
    # Windows
    r'\bdel\s+/[sq]', r'\brd\s+/s', r'\brmdir\s+/s',
    r'\bformat\s+[a-zA-Z]:', r'\bdiskpart\b',
    r'Remove-Item.*-Recurse.*-Force', r'Stop-Process.*-Force',
    r'Clear-Content.*-Force', r'Set-ExecutionPolicy\s+Unrestricted',
]

DANGEROUS_RE = [re.compile(p, re.IGNORECASE) for p in DANGEROUS_PATTERNS]


def is_dangerous(tool_name: str, input: dict) -> bool:
    """Check if a tool call matches dangerous patterns."""
    if tool_name == 'exec':
        cmd = input.get('command', '')
        return any(r.search(cmd) for r in DANGEROUS_RE)
    if tool_name == 'write_file':
        path = input.get('path', '')
        # Writing to system paths
        if path.startswith('/etc/') or path.startswith('/usr/') or path.startswith('/bin/'):
            return True
    return False


def summarize(tool_name: str, input: dict) -> str:
    """Human-readable summary of a tool call."""
    if tool_name == 'exec':
        return input.get('command', '')[:120]
    if tool_name in ('write_file', 'edit_file', 'read_file'):
        return input.get('path', '')
    if tool_name == 'web_fetch':
        return input.get('url', '')[:100]
    if tool_name == 'web_serve':
        return input.get('dir', input.get('directory', ''))[:80]
    return str(input)[:80]


def make_rule(tool_name: str, input: dict) -> str:
    """Generate a session allow-rule from a tool call.
    e.g. exec:git* for 'git status', write_file:src/* for 'src/app.py'
    """
    if tool_name == 'exec':
        cmd = input.get('command', '').strip()
        first_word = cmd.split()[0] if cmd else ''
        if '/' in first_word:
            first_word = first_word.rsplit('/', 1)[-1]
        return f'exec:{first_word}*' if first_word else 'exec:*'
    if tool_name in ('write_file', 'edit_file'):
        path = input.get('path', '')
        if '/' in path:
            dir_part = path.rsplit('/', 1)[0]
            return f'{tool_name}:{dir_part}/*'
        return f'{tool_name}:*'
    return f'{tool_name}:*'


def matches_rule(rule: str, tool_name: str, input: dict) -> bool:
    """Check if a tool call matches an allow rule."""
    if ':' not in rule:
        return rule == tool_name
    rule_tool, rule_pattern = rule.split(':', 1)
    if rule_tool != tool_name:
        return False
    if rule_pattern == '*':
        return True

    if tool_name == 'exec':
        cmd = input.get('command', '').strip()
        # Pattern like "git*" matches "git status", "git push", etc.
        if rule_pattern.endswith('*'):
            prefix = rule_pattern[:-1]
            return cmd.startswith(prefix) or cmd.split()[0] == prefix.rstrip()
        return cmd.startswith(rule_pattern)
    if tool_name in ('write_file', 'edit_file'):
        path = input.get('path', '')
        if rule_pattern.endswith('/*'):
            dir_prefix = rule_pattern[:-2]
            return path.startswith(dir_prefix + '/')
        return path == rule_pattern
    return False


class TuiPermissions:
    """Permission system for the TUI with three modes:

    - auto: approve everything except dangerous commands
    - ask: prompt for every non-safe tool, with option to add session rules
    - locked: approve nothing that isn't in ALWAYS_SAFE
    """

    MODES = ('ask', 'auto', 'locked', 'yolo')

    def __init__(self, app=None, renderer=None):
        self.app = app
        self.renderer = renderer  # for one-shot mode fallback
        self.mode = 'auto' if app is None else 'ask'
        self.session_rules = set()
        self.approve_all = False

    def is_auto_approved(self, tool_name: str, input: dict) -> bool:
        if tool_name in ALWAYS_SAFE:
            return True
        if self.mode == 'yolo':
            return True  # everything approved, no exceptions
        if self.approve_all or self.mode == 'auto':
            if is_dangerous(tool_name, input):
                return False  # dangerous always asks
            return True
        if self.mode == 'locked':
            return False
        # ask mode — check session rules
        for rule in self.session_rules:
            if matches_rule(rule, tool_name, input):
                return True
        return False

    def _notify_companion(self, tool_name, summary, dangerous):
        """Send tool:awaiting-approval to companion app."""
        try:
            import json, asyncio
            conn = self.app.conn
            if conn and conn.connected:
                asyncio.ensure_future(conn.send(json.dumps({
                    'type': 'tool:awaiting-approval',
                    'name': tool_name,
                    'summary': summary,
                    'dangerous': dangerous,
                })))
        except Exception:
            pass

    async def prompt(self, tool_name: str, input: dict) -> bool:
        """Show approval UI. Uses TUI selector if app is available, console prompt otherwise."""
        summary = summarize(tool_name, input)
        rule = make_rule(tool_name, input)

        # One-shot / non-TUI mode — simple console prompt
        if not self.app:
            from rich.prompt import Confirm
            loop = __import__('asyncio').get_event_loop()
            return await loop.run_in_executor(
                None, lambda: Confirm.ask(f'  Allow [bold]{tool_name}[/bold]: {summary}?', default=True)
            )
        dangerous = is_dangerous(tool_name, input)

        from rich.text import Text
        t = self.app.theme_data

        # Show what's being requested
        label = Text()
        label.append(f'  ⚙ {tool_name}: ', style=f'bold {t["accent"]}')
        label.append(summary, style=t['fg'])
        if dangerous:
            label.append('  ⚠ DANGEROUS', style=t['error'])
        self.app._log(label)
        self.app._scroll_bottom()

        # Build options
        if dangerous:
            options = ['✓ Allow (once)', '✗ Deny']
        else:
            options = ['✓ Allow', f'✓ Allow all {rule}', '✗ Deny']

        # Use PromptProvider — serialized with lock, one prompt at a time.
        # Notify companion app right before showing the prompt (inside the lock
        # ensures only one notification at a time, matching one prompt at a time).
        result = await self.app.prompter.choice(
            f'Allow {tool_name}?', options,
            on_show=lambda: self._notify_companion(tool_name, summary, dangerous),
        )

        if result.get('cancelled'):
            return False

        choice = result.get('index', -1)
        if dangerous:
            allowed = (choice == 0)
        else:
            allowed = (choice in (0, 1))
            if choice == 1 and rule:
                self.session_rules.add(rule)
                self.app._log(Text(f'  ✓ Rule added for session: {rule}', style=t['success']))

        if allowed:
            self.app._log(Text(f'  ✓ Allowed', style=t['success']))
        else:
            self.app._log(Text(f'  ✗ Denied', style=t.get('warning', 'yellow')))
        self.app._scroll_bottom()

        # Broadcast to companion app so approval cards dismiss
        try:
            self.app.bridge.broadcast('interactive:resolved', kind='tool-approval', allowed=allowed)
        except Exception:
            pass

        return allowed

File: internal/app/permissions.go
package app

import (
	"fmt"
	"regexp"
	"strings"
	"sync"
)

// PermMode maps to Python's TuiPermissions modes.
type PermMode string

const (
	PermAuto   PermMode = "auto"
	PermAsk    PermMode = "ask"
	PermLocked PermMode = "locked"
	PermYolo   PermMode = "yolo"
)

// alwaysSafe — never need approval.
var alwaysSafe = map[string]bool{
	"read_file": true, "glob": true, "grep": true,
}

// dangerousPatterns — require approval even in auto. Mirrors
// acorn/permissions.py:DANGEROUS_PATTERNS.
var dangerousPatternsRe = compileAll([]string{
	`\brm\s+(-rf?|--recursive)`, `\brm\s+/`, `rmdir\s+/`,
	`\bmkfs\b`, `>\s*/dev/`, `dd\s+if=`,
	`chmod\s+(-R\s+)?777`, `chown\s+-R\s+.*/`,
	`\bgit\s+push\s+.*--force`, `\bgit\s+reset\s+--hard`,
	`\bdrop\s+table\b`, `\bdrop\s+database\b`,
	`\btruncate\s+table\b`,
	`\bmkfs\.\w+\b`, `\bfdisk\b`, `\bparted\b`,
	`:\(\)\{`, `curl.*\|\s*(ba)?sh`, `wget.*\|\s*(ba)?sh`,
	`\bkill\s+-9\b`,
	`\bdel\s+/[sq]`, `\brd\s+/s`, `\brmdir\s+/s`,
	`\bformat\s+[a-zA-Z]:`, `\bdiskpart\b`,
	`Remove-Item.*-Recurse.*-Force`, `Stop-Process.*-Force`,
	`Clear-Content.*-Force`, `Set-ExecutionPolicy\s+Unrestricted`,
})

func compileAll(ps []string) []*regexp.Regexp {
	out := make([]*regexp.Regexp, 0, len(ps))
	for _, p := range ps {
		out = append(out, regexp.MustCompile("(?i)"+p))
	}
	return out
}

// IsDangerous checks whether a tool call matches dangerous patterns.
func IsDangerous(name string, input map[string]any) bool {
	if name == "exec" {
		cmd, _ := input["command"].(string)
		for _, r := range dangerousPatternsRe {
			if r.FindStringIndex(cmd) != nil {
				return true
			}
		}
	}
	if name == "write_file" {
		p, _ := input["path"].(string)
		if strings.HasPrefix(p, "/etc/") || strings.HasPrefix(p, "/usr/") || strings.HasPrefix(p, "/bin/") {
			return true
		}
		lp := strings.ReplaceAll(strings.ToLower(p), "/", `\`)
		for _, w := range []string{`c:\windows`, `c:\program files`, `c:\programdata`} {
			if strings.HasPrefix(lp, w) {
				return true
			}
		}
	}
	return false
}

// Summarize gives a one-line human label for a tool call.
func Summarize(name string, input map[string]any) string {
	switch name {
	case "exec":
		c, _ := input["command"].(string)
		if len(c) > 120 {
			c = c[:120] + "…"
		}
		return c
	case "write_file", "edit_file", "read_file":
		p, _ := input["path"].(string)
		return p
	case "web_fetch":
		u, _ := input["url"].(string)
		if len(u) > 100 {
			u = u[:100]
		}
		return u
	case "web_serve":
		if v, ok := input["dir"].(string); ok && v != "" {
			return v
		}
		if v, ok := input["directory"].(string); ok && v != "" {
			return v
		}
		return ""
	}
	return fmt.Sprintf("%v", input)
}

// MakeRule derives a session allow-rule from a tool call. Mirrors Python.
func MakeRule(name string, input map[string]any) string {
	switch name {
	case "exec":
		cmd, _ := input["command"].(string)
		cmd = strings.TrimSpace(cmd)
		parts := strings.Fields(cmd)
		first := ""
		if len(parts) > 0 {
			first = parts[0]
			if i := strings.LastIndex(first, "/"); i >= 0 {
				first = first[i+1:]
			}
		}
		if first == "" {
			return "exec:*"
		}
		return "exec:" + first + "*"
	case "write_file", "edit_file":
		p, _ := input["path"].(string)
		if i := strings.LastIndex(p, "/"); i >= 0 {
			return name + ":" + p[:i] + "/*"
		}
		return name + ":*"
	}
	return name + ":*"
}

// MatchesRule — true if the rule covers this tool call.
func MatchesRule(rule, name string, input map[string]any) bool {
	colon := strings.Index(rule, ":")
	if colon < 0 {
		return rule == name
	}
	if rule[:colon] != name {
		return false
	}
	pat := rule[colon+1:]
	if pat == "*" {
		return true
	}
	switch name {
	case "exec":
		cmd, _ := input["command"].(string)
		cmd = strings.TrimSpace(cmd)
		if strings.HasSuffix(pat, "*") {
			prefix := strings.TrimSuffix(pat, "*")
			if strings.HasPrefix(cmd, prefix) {
				return true
			}
			first := ""
			if parts := strings.Fields(cmd); len(parts) > 0 {
				first = parts[0]
			}
			return first == strings.TrimRight(prefix, " ")
		}
		return strings.HasPrefix(cmd, pat)
	case "write_file", "edit_file":
		p, _ := input["path"].(string)
		if strings.HasSuffix(pat, "/*") {
			dir := strings.TrimSuffix(pat, "/*")
			return strings.HasPrefix(p, dir+"/")
		}
		return strings.HasPrefix(p, pat)
	}
	return false
}

// TUIPerms implements tools.Permissions backed by a modal on the Model.
// Approval goes through a blocking channel so we can open the modal on the
// UI thread and wait for the keystroke result.
type TUIPerms struct {
	m    *Model
	mu   sync.Mutex
	mode PermMode
	// Session-scope allow rules.
	rules []string
	// Pending prompt — a goroutine waits on ch for the UI to resolve.
	pendingCh chan bool
	pendingName  string
	pendingInput map[string]any
	pendingRule  string
}

func newTUIPerms(m *Model) *TUIPerms {
	return &TUIPerms{m: m, mode: PermAsk}
}

// SetMode switches the overall approval posture.
func (p *TUIPerms) SetMode(mode PermMode) {
	p.mu.Lock()
	p.mode = mode
	p.mu.Unlock()
}

// Mode returns the current permission mode.
func (p *TUIPerms) Mode() PermMode {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.mode
}

// AddRule grows the session allow list (e.g. "exec:git*").
func (p *TUIPerms) AddRule(r string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, e := range p.rules {
		if e == r {
			return
		}
	}
	p.rules = append(p.rules, r)
}

// Rules returns a copy of the current allow list.
func (p *TUIPerms) Rules() []string {
	p.mu.Lock()
	defer p.mu.Unlock()
	out := make([]string, len(p.rules))
	copy(out, p.rules)
	return out
}

// IsAutoApproved returns true if the tool call doesn't need a prompt.
func (p *TUIPerms) IsAutoApproved(name string, input map[string]any) bool {
	p.mu.Lock()
	mode := p.mode
	rules := append([]string(nil), p.rules...)
	p.mu.Unlock()
	if alwaysSafe[name] {
		return true
	}
	if mode == PermLocked {
		return false
	}
	if mode == PermYolo {
		return true
	}
	if mode == PermAuto {
		if IsDangerous(name, input) {
			return false
		}
		return true
	}
	// PermAsk — check session rules.
	for _, r := range rules {
		if MatchesRule(r, name, input) {
			return true
		}
	}
	return false
}

// Prompt opens the permission modal and blocks until the user decides.
// Runs off the Tea goroutine — the UI signals result via m.resolvePerm().
func (p *TUIPerms) Prompt(name string, input map[string]any) bool {
	// In yolo mode, never prompt. In locked mode, always deny without asking.
	p.mu.Lock()
	mode := p.mode
	p.mu.Unlock()
	if mode == PermYolo {
		return true
	}
	if mode == PermLocked {
		return false
	}
	ch := make(chan bool, 1)
	p.mu.Lock()
	p.pendingCh = ch
	p.pendingName = name
	p.pendingInput = input
	p.pendingRule = MakeRule(name, input)
	p.mu.Unlock()

	// Signal the Tea program to open the modal.
	if p.m != nil && p.m.sendProgramMsg != nil {
		p.m.sendProgramMsg(openPermModalMsg{
			name:    name,
			summary: Summarize(name, input),
			dangerous: IsDangerous(name, input),
			rule:    p.pendingRule,
		})
	}

	// Broadcast to observers (mobile / VS Code companion) so they can
	// render the same modal or at least show "operator approval needed"
	// state. Mirrors acorn/permissions.py:142. Safe off the Tea
	// goroutine: m.Broadcast → conn.Client.Send is mutex-locked.
	if p.m != nil {
		p.m.Broadcast("tool:awaiting-approval", map[string]any{
			"tool":      name,
			"summary":   Summarize(name, input),
			"dangerous": IsDangerous(name, input),
			"rule":      p.pendingRule,
		})
	}

	allowed := <-ch
	p.mu.Lock()
	p.pendingCh = nil
	p.pendingName = ""
	p.pendingInput = nil
	p.pendingRule = ""
	p.mu.Unlock()
	// Companion gets the resolution too so it can dismiss its prompt UI.
	if p.m != nil {
		p.m.Broadcast("tool:approval-resolved", map[string]any{
			"tool":    name,
			"allowed": allowed,
		})
	}
	return allowed
}

// resolvePerm is called from the UI thread to answer the pending prompt.
func (p *TUIPerms) resolvePerm(allowed bool, addRule bool) {
	p.mu.Lock()
	ch := p.pendingCh
	rule := p.pendingRule
	p.mu.Unlock()
	if ch == nil {
		return
	}
	if allowed && addRule && rule != "" {
		p.AddRule(rule)
	}
	select {
	case ch <- allowed:
	default:
	}
}

// openPermModalMsg is delivered to the Bubble Tea program when the
// permissions layer wants to open a modal. The UI thread picks it up,
// flips m.modal = modalPermission, stashes the data, and re-renders.
type openPermModalMsg struct {
	name, summary, rule string
	dangerous           bool
}
